File: backend/src/main/java/project/backend/global/security/handler/JwtAuthenticationFilter.java
package project.backend.global.security.handler;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import project.backend.global.exception.errorcode.TokenErrorCode;
import project.backend.global.exception.ex.JwtException;
import project.backend.global.security.app.CookieUtils;
import project.backend.global.security.jwt.JwtProvider;
import project.backend.global.security.jwt.TokenStatus;

@Slf4j
@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

	private final JwtProvider jwtProvider;

	private static final List<String> WHITE_LIST = List.of(
		"/signup",
		"/login"
	);

	@Override
	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
		FilterChain filterChain)
		throws ServletException, IOException {

		if (WHITE_LIST.contains(request.getRequestURI())) {
			filterChain.doFilter(request, response); // JWT 검사 건너뜀
			return;
		}

		log.info("JWT필터 도달 = {}", request.getRequestURI());

		Cookie cookie = CookieUtils.getCookie(request,
			"accessToken").orElseThrow(() -> new JwtException(TokenErrorCode.INVALID_TOKEN));
		String accessToken = cookie.getValue();

		TokenStatus tokenStatus = jwtProvider.validateAccessToken(accessToken);

		if (tokenStatus == TokenStatus.VALID) {
			log.info("[JWT] 유효한 토큰");
			Authentication authentication = jwtProvider.getAuthentication(accessToken);
			SecurityContextHolder.getContext().setAuthentication(authentication);

		} else {
			SecurityContextHolder.clearContext();
		}

		filterChain.doFilter(request, response);
	}

}

File: backend/src/main/java/project/backend/global/security/filter/JwtAuthenticationFilter.kt
package project.backend.global.security.filter

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import project.backend.auth.app.CookieUtils
import project.backend.auth.token.jwt.JwtProvider
import project.backend.auth.token.jwt.TokenStatus
import project.backend.global.exception.errorcode.TokenErrorCode
import project.backend.global.exception.ex.CustomJwtException

@Component
class JwtAuthenticationFilter(
    private val jwtProvider: JwtProvider
) : OncePerRequestFilter() {

    private val log = KotlinLogging.logger {}

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val requestURI = request.requestURI
        if (requestURI.startsWith("/github/") || requestURI in WHITE_LIST) {
            filterChain.doFilter(request, response) // JWT 검사 건너뜀
            return
        }

        log.info { "JWT 필터 도달 = ${request.requestURI}" }

        val accessToken = CookieUtils.getCookie(request, "accessToken")
            ?.value ?: throw CustomJwtException(TokenErrorCode.NOT_FOUND_TOKEN)

        when (jwtProvider.validateAccessToken(accessToken)) {
            TokenStatus.VALID -> {
                log.info { "[JWT] 유효한 토큰" }
                val authentication = jwtProvider.getAuthentication(accessToken)
                SecurityContextHolder.getContext().authentication = authentication
            }

            else -> {
                SecurityContextHolder.clearContext()
            }
        }

        filterChain.doFilter(request, response)
    }

    companion object {
        private val WHITE_LIST = listOf(
            "/signup",
            "/login",
            "/token/refresh"
        )
    }
}
